### PR TITLE
feat: add match-based rules for targeted directory validation

### PR DESCRIPTION
**Pull request by dallenpyrah**
## Problem

Defining the entire filesystem layout tree is impractical for large monorepos. You often only care about structure in specific directories (e.g., API modules), not everywhere.

## Solution

Add **match-based rules** that allow targeting specific directory patterns and enforcing requirements only there.

### Basic Example

```yaml
rules:
  match:
    - pattern: "apps/*/api/src/modules/*"
      require: [controller.ts, service.ts, repo.ts]
      allow: [errors.ts, lib]
      strict: true
      case: kebab        # validate directory name
      childCase: kebab   # validate children names
```

### All Options

| Option | Description |
|--------|-------------|
| `pattern` | Glob pattern to match directories |
| `exclude` | Patterns to exclude from matching |
| `require` | Required files/directories that must exist |
| `allow` | Allowed entries (used with strict mode) |
| `forbid` | Forbidden files/directories |
| `strict` | Only required + allowed entries permitted |
| `case` | Naming convention for matched directory itself |
| `childCase` | Naming convention for children of matched directories |

### Performance: O(n) instead of O(n²)
- Build directory tree once, traverse efficiently
- Previous implementation iterated all files for each directory check

### Bug Fixes
- **Empty allowedPatterns bug**: `strict: true` with no `require`/`allow` now rejects ALL entries (was allowing everything)
- **Path construction**: Properly handles root directory paths
- **Magic strings**: Added `Match` to `RuleNames` constant

### New Features
- `childCase` option for validating children separately from directory name
- `addWarning` helper for non-critical issues
- `inspect rule match` command support
- Warning when pattern matches nothing (likely config typo)

### Test Coverage: 19 New Tests

| Test Case | Status |
|-----------|--------|
| Empty directories | ✅ |
| Root-level matches | ✅ |
| Overlapping rules | ✅ |
| Pattern matches nothing (warning) | ✅ |
| Deeply nested structures (5+ levels) | ✅ |
| require and forbid with same entry | ✅ |
| Hidden files with case validation | ✅ |
| Files without extensions | ✅ |
| Glob patterns in require/allow/forbid | ✅ |
| strict mode with empty require/allow | ✅ |
| case vs childCase distinction | ✅ |

### Documentation
- Comprehensive README section with examples
- All options documented
- Behavior notes and edge cases explained

## Testing

```
162 pass
0 fail
314 expect() calls
Ran 162 tests across 15 files
```

### CODE REVIEW

File: src/rules/match.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      }
+    }
+
+    // Check strict mode (only required + allowed entries permitted)
+    if (rule.strict) {
+      const allowedPatterns = [...(rule.require ?? []), ...(rule.allow ?? [])];
+
+      for (const entry of allEntries) {
+        const isAllowed = allowedPatterns.length === 0 || matchesAny(entry, allowedPatterns);
+        if (!isAllowed) {
```
[the remark this answers is not in the document]
Comment (dallenpyrah) [reply src/rules/match.ts]: ✅ Fixed in commit b1aa912 - empty `allowedPatterns` now rejects ALL entries:

```typescript
if (allowedPatterns.length === 0) {
  for (const entry of allEntries) {
    yield* addViolation(ctx, {
      path: joinPath(node.path, entry),
      rule: RuleNames.Match,
      message: "entry not allowed (strict mode with no allowed patterns)",
    });
  }
}
```

Test coverage added: `test("strict mode with empty require/allow rejects all entries", ...)`
